op-program/client/interop/consolidate.go

```diff
@@ -0,0 +1,245 @@
+		hashByNum[chain.ChainID][head.NumberU64()] = head.Hash()
+	}
+	depset, err := depset.NewStaticConfigDependencySet(deps)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: failed to create dependency set: %w", err)
+	}
+
+	return &consolidateCheckDeps{
+		oracle:    oracle,
+		heads:     heads,
+		depset:    depset,
+		hashByNum: hashByNum,
+	}, nil
+}
+
+func (d *consolidateCheckDeps) Check(
+	chain eth.ChainID,
+	blockNum uint64,
+	timestamp uint64,
+	logIdx uint32,
+	logHash common.Hash,
+) (includedIn supervisortypes.BlockSeal, err error) {
+	// We can assume the oracle has the block the executing message is in
+	block, err := d.BlockByNumber(d.oracle, blockNum, chain.ToBig().Uint64())
+	if err != nil {
+		return supervisortypes.BlockSeal{}, err
+	}
+	return supervisortypes.BlockSeal{
+		Hash:      block.Hash(),
+		Number:    block.NumberU64(),
+		Timestamp: block.Time(),
+	}, nil
+}
+
+func (d *consolidateCheckDeps) IsCrossUnsafe(chainID eth.ChainID, block eth.BlockID) error {
+	// Assumed to be cross-unsafe. And hazard checks will catch any future blocks prior to calling this
+	return nil
+}
+
+func (d *consolidateCheckDeps) IsLocalUnsafe(chainID eth.ChainID, block eth.BlockID) error {
+	// Always assumed to be local-unsafe
+	return nil
+}
+
+func (d *consolidateCheckDeps) ParentBlock(chainID eth.ChainID, parentOf eth.BlockID) (parent eth.BlockID, err error) {
+	block, err := d.BlockByNumber(d.oracle, parentOf.Number-1, chainID.ToBig().Uint64())
+	if err != nil {
+		return eth.BlockID{}, err
+	}
+	return eth.BlockID{
+		Hash:   block.Hash(),
+		Number: block.NumberU64(),
+	}, nil
+}
+
+func (d *consolidateCheckDeps) OpenBlock(
+	chainID eth.ChainID,
+	blockNum uint64,
+) (ref eth.BlockRef, logCount uint32, execMsgs map[uint32]*supervisortypes.ExecutingMessage, err error) {
+	block, err := d.BlockByNumber(d.oracle, blockNum, chainID.ToBig().Uint64())
+	if err != nil {
+		return eth.BlockRef{}, 0, nil, err
+	}
+	ref = eth.BlockRef{
+		Hash:   block.Hash(),
+		Number: block.NumberU64(),
+	}
+	_, receipts := d.oracle.ReceiptsByBlockHash(block.Hash(), chainID.ToBig().Uint64())
+	execs, logCount, err := ReceiptsToExecutingMessages(d.depset, receipts)
+	if err != nil {
+		return eth.BlockRef{}, 0, nil, err
+	}
+	execMsgs = make(map[uint32]*supervisortypes.ExecutingMessage, len(execs))
+	for _, exec := range execs {
+		execMsgs[exec.LogIdx] = exec
+	}
+	return ref, uint32(logCount), execMsgs, nil
+}
+
+func (d *consolidateCheckDeps) DependencySet() depset.DependencySet {
+	return d.depset
+}
+
+func (d *consolidateCheckDeps) BlockByNumber(oracle l2.Oracle, blockNum uint64, chainID uint64) (*ethtypes.Block, error) {
+	head := d.heads[chainID]
+	if head == nil {
+		return nil, fmt.Errorf("head not found for chain %v", chainID)
+	}
+	if head.NumberU64() < blockNum {
+		return nil, nil
+	}
+	hash, ok := d.hashByNum[chainID][blockNum]
+	if ok {
+		return oracle.BlockByHash(hash, chainID), nil
+	}
+	for head.NumberU64() > blockNum {
```

Comment:
This is going to walk back through the same blocks unnecessarily.  Imagine we lookup block number 10, so we walk back from our head at 20 to 10, caching the hashes along the way.  But then we get a request for block 5 which isn't in the cache - we'll start walking back from 20 all the way to 5 again. 

`OracleBackedL2Chain` has this logic right where it tracks the earliest indexed block so it can just start from there if it needs something earlier.  Would really like to avoid duplicating that code.

Reply:
`OracleBackedL2Chain` doesn't quite have the logic correct. I'll introduce a reusable construct that fixes both| uses.